keps/sig-node/20181106-in-place-update-of-pod-resources.md

```diff
@@ -200,6 +227,34 @@ Spec.Containers[i].ResourcesAllocated are used to determine if there is enough
 room to admit the Pod. Kubelet does not set Pod's ResourcesAllocated after
 admitting a new Pod.
 
+<<[UNRESOLVED]>>
+The above is confusing.  Does it mean that ResourcesAllocated will be set by
+someone other than Kubelet?  Who?  Why?
+
+Above you said 'limit access to ResourcesAllocated field such that only Kubelet
+can update this field' which is the opposite of 'kubelet does not set Pod's
+ResourcesAllocated after admitting' here.  I am very confused.
```

Comment:
This applies only to new pod admissions (HandlePodAdditions function in kubelet.go) - we don't set it in that case  to reduce the load on APIserver - we admit it at initial ResourcesAllocated levels - which the admission plugin sets equal to Resources.Requests it not set, or ensures it is equal as mentioned above.

Once admitted, updates to Resources.Requests for CPU/memory would result in kubelet patching the  ResourcesAllocated if node has enough resources to grant the new requests.

Reply:
I see.  This "optimization" feels superficial to me, but I admit I am not close to the details there.  Is it based on benchmarks or known scale issues?  E.g. is it a blanket rule that kubelet MUST NOT ever grow any feature that requires immediate writing to pod?  Is that well understood?

Reply:
It's more than that - today for every pod that is admitted, we don't have this trip to apiserver to set ResourcesAllocated or anything else in the podspec. Doing so now it would make it a regression, even though it is a clear signal of kubelet admitting the pod. It is in this context someone mentioned API ratelimiting in one of the sig meeting disucssions.

Besides, pod acceptance information is reflected in pod phase / pod conditions a little bit later. 

I feel it's OK to boldly document and enforce users not setting ResourcesAllocated. We can change it if the need arises - it can't be much worse than our change in this KEP making resources mutable.

Reply:
My point was that I didn't / don't know WHY we have this restriction.  Do we have some analysis that shows 1 additional API op per pod is a big deal?  Is it SUCH a big deal that we're willing to complicate the rest of the system to avoid it?  Where is that documented?  You can't call it a regression unless you have a critical benchmark you're tracking.  :)

In the bigger picture this may not be important, but it feels a bit like voodoo engineering.  We should do the thing that makes the most sense unless we have a very clear reason not to.

Reply:
I gave you the wrong impression in saying this was done to 'reduce load', when I actually meant 'so as to not create additional new load' during new pod addition.

That callout in the KEP was merely to clarify that we will set ResourcesAllocated when your new desired Request is acceptable to the node. But for new pod addition, we are not looking at requests anymore, we are looking at ResourcesAllocated instead and expect Resources==ResourcesAllocated. This lets kubelet handle restart scenario if Requests changed in the interim. It is not a restriction, we can set it and things will functionally be the same - but there is no need to. The current user-observed behavior continues to be unchanged.

Reply:
Point 1) I categorically reject arguments around "avoiding scalability problems" unless we have evidence that this is either actually a problem or very likely to be a problem (in which case I want to see analysis), especially when it forces the design into unnatural positions to do so.

Point 2) "This lets kubelet handle restart scenario if Requests changed in the interim".  IOW you're using the kube-apiserver as a checkpoint, and you're (IMO) adulterating the design in order to do so.  I have not seen a clear explanation of why the checkpoint (local or remote) is actually needed (I tried to work through it and I believe it is nice-to-have but not required), and even if so, I have not seen why it should be part of the user-facing API.

Sorry to be a pain in the butt - I want to understand the design so I can stand behind it.

Reply:
No worries, no apologies needed - I really appreciate the thorough review. We have to shake the table from different viewpoints to know it has strong legs. It is way too easy to get tunnel-visioned working with such large engineering codebase (talking about myself here). 

1. I'm not aware if this will cause scalability problems, there is no data that I'm aware of. But my point that ```* kubelet sets `allocated` = `resources` immediately (acts as an ACK)``` isn't necessary is valid. We are not doing it today when we add a new pod, why start doing it now? Extrapolating, if we start doing it, then conceivably we increase load on apiserver.
1. Yes, we are using podSpec as source-of-truth. Does above [scenario](https://github.com/kubernetes/enhancements/pull/1883#discussion_r449739475) with pods p1 & p2 in case 2 clarify the need for checkpointing? (either on apiserver or node-locally)

Reply:
About point 1, today we roughly have the following flow for new pods:
1. User/controller creates pod and posts it to apiserver
1. scheduler assigns a value to pod.Spec.NodeName to pod via /binding subresource
1. kubelet watching pods assigned to its node triggers pod admission check.
    - Focusing on resources, this boils down to: pod-requests <= node-available.
    - node-available = node-allocatable - sum(existing pod-requests)
1. Outcomes:
    - pod requests > node-available: kubelet sets PodPhase == Failed
    - pod requests <=  node-available: kubelet goes on to create pod sandbox, start containers, sets & update pod status as things progress


With our design, everything above remains the same except that we use pod resourcesAllocated (set equal to requests for new pod) instead of pod requests (which now mutable by user or VPA and can change at will). When we compute if a second new pod is admissible, we sum the resourcesAllocated of currently admitted pods instead of requests.

Why do we need to introduce the extra step of ACKing to apiserver that ```canAdmitPod``` internal function succeeded by writing back allocated = resources.requests?

Reply:
To (1): What do we expect in terms of pod starts per second across a large cluster?  I've heard numbers like O(1000) pods per second as a goal, but is that sustained?  Remember that most pods causes endpoints to be updated, events to be fired, etc.  I'm skeptical that one write is going to be the straw that breaks our back.

> Why do we need to introduce the extra step of ACKing to apiserver

I'm not saying we NEED to.  It was a side-effect of some other assumptions around the design, and a desire not to have users guessing about what is happening in the system.